nemo/lightning/pytorch/callbacks/megatron_comm_overlap.py

```diff
@@ -209,6 +209,32 @@ def _override_user_cfgs(self, comm_overlap_cfg):
 
         return comm_overlap_cfg
 
+    def _set_num_cuda_device_max_connections(self):
+        import os
+
+        import torch
+
+        from nemo.utils import AppState
+
+        app_state = AppState()
+        tp_size = app_state.tensor_model_parallel_size
+        cp_size = app_state.context_parallel_size
+        dp_size = app_state.data_parallel_size
+        pp_size = app_state.pipeline_model_parallel_size
+        major, _ = torch.cuda.get_device_capability()
+        if major > 9:
+            if (tp_size > 1 or cp_size > 1) and (dp_size > 1 or pp_size > 1):
+                # Default is 8, but for this case, we need extra connections
+                # to avoid serialization of streams
+                os.environ['CUDA_DEVICE_MAX_CONNECTIONS'] = "32"
+            else:
+                os.environ.pop('CUDA_DEVICE_MAX_CONNECTIONS')
+        else:
+            if tp_size > 1 or cp_size > 1:
+                os.environ['CUDA_DEVICE_MAX_CONNECTIONS'] = "1"
```

Comment:
It could also be good to add a doc string for this condition.
`Set the device connection to 1 to enforce the kernel queuing order from the host to the execution order on GPU. This is needed to schedule a communication kernel before the overlapping persistent GEMM kernel. Otherwise, the communication kernel will be pushed to the end of the GEMM kernel so failing to overlap the kernels`